tests/test_target.py

```diff
@@ -14,136 +14,158 @@
+                logger.info('> QEMU target runners:')
+                for entry in target_configs['QEMUTargetRunner'].values():
+                    logger.info('%s', repr(entry))
+
+                    qemu_runner = QEMUTargetRunner(
+                        qemu_settings=entry.get('qemu_settings'),
+                        connection_settings=entry.get('connection_settings'),
+                    )
+
+                    if entry.get('ChromeOsTarget') is not None:
+                        # Leave termination of QEMU runner to ChromeOS target.
+                        target_runners.append(NOPTargetRunner(qemu_runner.target))
+
+                        logger.info('>> ChromeOS target: %s', repr(entry["ChromeOsTarget"]))
+                        qemu_runner.target = ChromeOsTarget(
+                            connection_settings={
+                                **entry['ChromeOsTarget']['connection_settings'],
+                                **qemu_runner.target.connection_settings,
+                            },
+                            working_directory='/tmp/devlib-target',
+                        )
+
+                    target_runners.append(qemu_runner)
+
+            return target_runners
+
+        cls._target_runners = build_target_runners()
+
+    def test_read_multiline_values(self):
+        """
+        Test Target.read_tree_values_flat()
+
+        Runs tests around ``Target.read_tree_values_flat()`` for ``TargetRunner`` objects.
+        """
+
+        logger.info('Running test_read_multiline_values test...')
+
+        data = {
+            'test1': '1',
+            'test2': '2\n\n',
+            'test3': '3\n\n4\n\n',
+        }
+
+        for target_runner in self._target_runners:
+            target = target_runner.target
+
+            logger.info('target=%s os=%s hostname=%s',
+                        target.__class__.__name__, target.os, target.hostname)
+
+            with target.make_temp() as tempdir:
+                logger.debug('Created %s.', tempdir)
+
+                for key, value in data.items():
+                    path = os.path.join(tempdir, key)
+                    logger.debug('Writing %s to %s...', repr(value), path)
+                    target.write_value(path, value, verify=False,
+                                       as_root=target.conn.connected_as_root)
+
+                logger.debug('Reading values from target...')
+                raw_result = target.read_tree_values_flat(tempdir)
+                result = {os.path.basename(k): v for k, v in raw_result.items()}
+
+            assert {k: v.strip() for k, v in data.items()} == result
+
+    @classmethod
+    def tearDownClass(cls):
+        logger.info("Destroying resources...")
+
+        for target_runner in cls._target_runners:
+            target = target_runner.target
+
+            logger.debug('Removing %s...', target.working_directory)
+            target.remove(target.working_directory)
```

Comment:
In WA we have a few workload that benefit from keeping large files (photos, videos etc.) on the device between runs to reduce the setup costs/overhead. By default most of these workloads keep their files under the working_directory so it could be undesirable to wipe this directory on each disconnect. 

The general pattern we have tried to follow with our WA asset deployment mechanism is that if a files/directory is not needed after a given run is that we remove it but we still leave the option for workloads to override this as they choose.

Reply:
Maybe it would make sense to default to cleaning up, and have WA override that default then ? In general, keeping resources deployed on the target are a recipe for trouble unless carefully implemented (e.g. checksuming the resources to ensure what is already deployed is actually what the code expects to work with). Even if WA does that carefully to avoid troubles, I wouldn't expect most client code to go to these length. In LISA, we push all our tools every time they are needed (at most once per Target instance, so it's "cached" but only very temporarily).

Reply:
True, I agree it is safer in the general case to push to the target each time but even today a client can currently do so without issue. I would just question whether it makes sense for devlib to automatically handle the cleanup vs leave the decision to the client code itself via whatever mechanism they choose.
  
So far we haven't placed any restrictions on the expectations on the working directory, so I'd be hesitant to introduce a new default that would wipe it out as we can't say how other users / client code could be using this directory.

Essentially I think I might be missing what the benefit would be to the end user vs the potential downsides of clearing files the user may not have wanted to?

Reply:
I'll keep this hunk as is since this discussion sounds like a separate/future improvement.

Reply:
Let's move the discussion to https://github.com/ARM-software/devlib/issues/680, but let's keep that comment thread open as the code will need to be changed if we end up implementing the feature (or at least leave a `# TODO: revisit based on https://github.com/ARM-software/devlib/issues/680 outcome` in the code).

Reply:
Done.